plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs

```diff
@@ -416,14 +417,7 @@ unionVal (Value l) (Value r) =
 -- | Combine two 'Value' maps with the argument function.
 -- Assumes the well-definedness of the two maps.
 unionWith :: (Integer -> Integer -> Integer) -> Value -> Value -> Value
-unionWith f ls rs =
-    let
-        combined = unionVal ls rs
-        unThese k' = case k' of
-            This a    -> f a 0
-            That b    -> f 0 b
-            These a b -> f a b
-    in Value (fmap (fmap unThese) combined)
+unionWith f = coerce (Map.unionWith @CurrencySymbol (Map.unionWith @TokenName f))
```

Comment:
@zliu41 I don't really know much about `Value` stuff, is this actually correct?

Reply:
Yes, semantics is different. But isn't it correct _now_? It was quite weird before, I'd argue.

Reply:
Good point. It should be documented that it treats `[]` and `[(token, 0)]` as different Values (so does the old implementation, but in a slightly different way).

Reply:
So what is the expected behaviour of `unionWith`? What does it mean that this version is correct?

Reply:
Well, I by default expect the standard behavior, i.e. the [one](https://hackage.haskell.org/package/containers-0.7/docs/Data-Map-Internal.html#v:unionWith) in containers.

On the other hand, maybe I'm wrong, maybe the old behavior of `unionValue (-)` is the correct one. But in that case we should introduce

```haskell
unionWithDef :: forall k a. Eq k => a -> (a -> a -> a) -> Map k a -> Map k a -> Map k a
```

which feed the first argument into the second whenever a key is missing in one of the maps.

So yeah, I think you've convinced me that the old behavior was correct and my new one isn't. @zliu41 opinion?

Reply:
The old behavior doesn't treat a token with zero quantity and a token that doesn't exist the same way either. e.g.,

```
unionWith (\_ _ -> 42) [] [] == []
unionWith (\_ _ -> 42) [(token, 0)], [] == [(token, 42)]
```

This doesn't look any more correct than the new behavior.

> So what is the expected behaviour of unionWith? What does it mean that this version is correct?

I think we can document that the expected behavior is exactly how the new implementation behaves, and move on. I don't think this really makes a difference on anyone, since nobody is going to actually use something like `\_ _ -> 42` as the combining function.

Reply:
If possible, it's also good to add a test which demonstrates this corner case.

Reply:
Yes some tests would be nice.